crypto/evp/keymgmt_lib.c

```diff
@@ -211,3 +225,260 @@ void *evp_keymgmt_util_fromdata(EVP_PKEY *target, EVP_KEYMGMT *keymgmt,
 
     return keydata;
 }
+
+int evp_keymgmt_util_has(EVP_PKEY *pk, int selection)
+{
+    size_t i, end = OSSL_NELEM(pk->pkeys);
+
+    for (i = 0; i < end && pk->pkeys[i].keymgmt != NULL; i++) {
+        const EVP_KEYMGMT *keymgmt = pk->pkeys[i].keymgmt;
+        void *keydata = pk->pkeys[i].keydata;
+
+        if (keymgmt->has != NULL)
+            return evp_keymgmt_has(keymgmt, keydata, selection);
+    }
+    /*
+     * TODO(3.0) investigate whether the lack of any |has| function should
+     * mean that the key has the components indicated by |selection| (i.e.
+     * they are assumed to be the empty set) or not.
+     * TODO(3.0) Investigate if we should return a value that indicates that
+     * this function is unsupported, leaving it to the caller to decide what
+     * to do.
+     *
+     * We currently assume that if there is no |has| function, the components
+     * indicated by |selection| are present, because that's how
+     * EVP_PKEY_missing_parameters() behaves for legacy EVP_PKEYs.
+     */
+    return 1;
+}
+
+/*
+ * Specialized function to find matching keymgmt in two EVP_PKEYs, given
+ * the return value from a |finder| function.  This function assumes that
+ * if there isn't a perfect match, the caller will have to export one of
+ * the keys to the other key's importing keymgmt, so it looks for that
+ * too.
+ *
+ * No return value.  The success is determined by the values assigned to
+ * the pointer references.
+ */
+static void find_keymgmt(EVP_KEYMGMT **keymgmt1, EVP_KEYMGMT **keymgmt2,
+                         void **keydata1, void **keydata2,
+                         EVP_PKEY *pk1, EVP_PKEY *pk2,
+                         int (*finder)(EVP_KEYMGMT *keymgmt))
+{
+    EVP_KEYMGMT *tmp_keymgmt1 = NULL, *tmp_keymgmt2 = NULL;
+    void *tmp_keydata1 = NULL, *tmp_keydata2 = NULL;
+    size_t i1, i2, end = OSSL_NELEM(pk1->pkeys);
+
+    /*
+     * Find cache elements that share the same keymgmt, which must also have
+     * function determined by |finder|.  This search is obviously O(n^2), but
+     * since the cache is fairly small, we assume that it's still cheaper
+     * than a export + import.
+     */
+    for (i1 = 0; i1 < end && pk1->pkeys[i1].keymgmt != NULL; i1++) {
+        if (finder(pk1->pkeys[i1].keymgmt)) {
+            tmp_keymgmt1 = pk1->pkeys[i1].keymgmt;
+            tmp_keydata1 = pk1->pkeys[i1].keydata;
+        }
+
+        for (i2 = 0; i2 < end && pk2->pkeys[i2].keymgmt != NULL; i2++) {
```

Comment:
I think you need a end2 = OSSL_NELEM(pk2->pkeys)..

Reply:
"need" might be overstating it, or do you expect `end1` and `end2` to differ much?

I do agree that it would be more elegant, though, so will fix

Reply:
I dont know will be in each one currently - so I would not like to guess that they are the same size :)

Reply:
Yeahok.  I've added the change, I'm sure the result will amuse you :wink: